File: apps/mobile/modules/domovoi-device-key/android/src/main/java/expo/modules/domovoidevicekey/DomovoiDeviceKeyModule.kt
package expo.modules.domovoidevicekey

import android.os.Build
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyInfo
import android.security.keystore.KeyProperties
import android.util.Base64
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.math.BigInteger
import java.security.AlgorithmParameters
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.KeyStore
import java.security.PrivateKey
import java.security.PublicKey
import java.security.SecureRandom
import java.security.interfaces.ECPublicKey
import java.security.spec.ECGenParameterSpec
import java.security.spec.ECParameterSpec
import java.security.spec.ECPoint
import java.security.spec.ECPublicKeySpec
import javax.crypto.KeyAgreement

private const val KEYSTORE = "AndroidKeyStore"
private const val CURVE = "secp256r1"
private const val COORDINATE_BYTES = 32
private const val UNCOMPRESSED_POINT_BYTES = 65

class DeviceKeyUnsupportedException(message: String) :
  CodedException("ERR_DEVICE_KEY_UNSUPPORTED", message, null)

class DeviceKeyMissingException(alias: String) :
  CodedException("ERR_DEVICE_KEY_MISSING", "No static key is stored under $alias", null)

class DeviceKeyPeerException(message: String) :
  CodedException("ERR_DEVICE_KEY_PEER", message, null)

class DomovoiDeviceKeyModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("DomovoiDeviceKey")

    Function("isSupported") { Build.VERSION.SDK_INT >= Build.VERSION_CODES.S }

    AsyncFunction("createStaticKey") { alias: String, requireHardware: Boolean ->
      requireAgreementSupport()
      val store = openStore()
      if (store.containsAlias(alias)) store.deleteEntry(alias)
      generate(alias, strongBox = true)
      val report = report(alias)
      if (requireHardware && report["securityLevel"] == "software") {
        openStore().deleteEntry(alias)
        throw DeviceKeyUnsupportedException("This device generated the static key in software")
      }
      report
    }

    AsyncFunction("getStaticKey") { alias: String ->
      if (!openStore().containsAlias(alias)) null else report(alias)
    }

    AsyncFunction("agree") { alias: String, peerPublicKey: String ->
      requireAgreementSupport()
      val privateKey = openStore().getKey(alias, null) as? PrivateKey
        ?: throw DeviceKeyMissingException(alias)
      val agreement = KeyAgreement.getInstance("ECDH", KEYSTORE)
      agreement.init(privateKey)
      agreement.doPhase(decodePeer(peerPublicKey), true)
      Base64.encodeToString(agreement.generateSecret(), Base64.NO_WRAP)
    }

    AsyncFunction("deleteStaticKey") { alias: String ->
      val store = openStore()
      val existed = store.containsAlias(alias)
      if (existed) store.deleteEntry(alias)
      existed
    }

    Function("randomBytes") { count: Int ->
      if (count <= 0 || count > 1024) throw DeviceKeyPeerException("Ask for 1 to 1024 bytes")
      val bytes = ByteArray(count)
      SecureRandom().nextBytes(bytes)
      Base64.encodeToString(bytes, Base64.NO_WRAP)
    }
  }

  private fun openStore(): KeyStore = KeyStore.getInstance(KEYSTORE).apply { load(null) }

  private fun requireAgreementSupport() {
    // PURPOSE_AGREE_KEY, and so a non-exportable P-256 agreement key, starts at API 31.
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) {
      throw DeviceKeyUnsupportedException("Key agreement in the platform key store needs Android 12")
    }
  }

  private fun generate(alias: String, strongBox: Boolean) {
    val generator = KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_EC, KEYSTORE)
    val spec = KeyGenParameterSpec.Builder(alias, KeyProperties.PURPOSE_AGREE_KEY)
      .setAlgorithmParameterSpec(ECGenParameterSpec(CURVE))
      .setDigests(KeyProperties.DIGEST_SHA256)
      .also { if (strongBox) it.setIsStrongBoxBacked(true) }
      .build()
    try {
      generator.initialize(spec)
      generator.generateKeyPair()
    } catch (error: Exception) {
      // A device without StrongBox refuses the request rather than quietly downgrading it.
      if (!strongBox) throw error
      generate(alias, strongBox = false)
    }
  }

  private fun report(alias: String): Map<String, Any> {
    val store = openStore()
    val privateKey = store.getKey(alias, null) as? PrivateKey
      ?: throw DeviceKeyMissingException(alias)
    val certificate = store.getCertificate(alias) ?: throw DeviceKeyMissingException(alias)
    val point = (certificate.publicKey as ECPublicKey).w
    val encoded = ByteArray(UNCOMPRESSED_POINT_BYTES)
    encoded[0] = 4
    writeCoordinate(point.affineX, encoded, 1)
    writeCoordinate(point.affineY, encoded, 1 + COORDINATE_BYTES)
    val info = KeyFactory.getInstance(privateKey.algorithm, KEYSTORE)
      .getKeySpec(privateKey, KeyInfo::class.java)
    return mapOf(
      "publicKey" to Base64.encodeToString(encoded, Base64.NO_WRAP),
      "securityLevel" to securityLevel(info),
    )
  }

  private fun securityLevel(info: KeyInfo): String {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
      return when (info.securityLevel) {
        KeyProperties.SECURITY_LEVEL_STRONGBOX -> "strongbox"
        KeyProperties.SECURITY_LEVEL_TRUSTED_ENVIRONMENT -> "trusted-environment"
        KeyProperties.SECURITY_LEVEL_SOFTWARE -> "software"
        else -> "unknown"
      }
    }
    @Suppress("DEPRECATION")
    return if (info.isInsideSecureHardware) "trusted-environment" else "software"
  }

  private fun writeCoordinate(coordinate: BigInteger, target: ByteArray, offset: Int) {
    // BigInteger.toByteArray is signed and variable length; the wire format is a fixed 32 bytes.
    val source = coordinate.toByteArray()
    val trimmed = if (source.size > COORDINATE_BYTES) {
      source.copyOfRange(source.size - COORDINATE_BYTES, source.size)
    } else {
      source
    }
    System.arraycopy(trimmed, 0, target, offset + (COORDINATE_BYTES - trimmed.size), trimmed.size)
  }

  private fun decodePeer(peerPublicKey: String): PublicKey {
    val bytes = try {
      Base64.decode(peerPublicKey, Base64.NO_WRAP)
    } catch (error: IllegalArgumentException) {
      throw DeviceKeyPeerException("The peer public key is not base64")
    }
    if (bytes.size != UNCOMPRESSED_POINT_BYTES || bytes[0].toInt() != 4) {
      throw DeviceKeyPeerException("The peer public key must be a 65 byte uncompressed P-256 point")
    }
    val x = BigInteger(1, bytes.copyOfRange(1, 1 + COORDINATE_BYTES))
    val y = BigInteger(1, bytes.copyOfRange(1 + COORDINATE_BYTES, UNCOMPRESSED_POINT_BYTES))
    val parameters = AlgorithmParameters.getInstance("EC").apply { init(ECGenParameterSpec(CURVE)) }
    val spec = parameters.getParameterSpec(ECParameterSpec::class.java)
    return KeyFactory.getInstance("EC").generatePublic(ECPublicKeySpec(ECPoint(x, y), spec))
  }
}

File: apps/mobile/modules/domovoi-device-key/ios/DomovoiDeviceKeyModule.swift
import CryptoKit
import ExpoModulesCore
import Security

private let service = "sh.domovoi.device-key"
private let uncompressedPointBytes = 65

private func keychainQuery(_ alias: String) -> [String: Any] {
  [
    kSecClass as String: kSecClassGenericPassword,
    kSecAttrService as String: service,
    kSecAttrAccount as String: alias,
  ]
}

private func loadRepresentation(_ alias: String) -> Data? {
  var query = keychainQuery(alias)
  query[kSecReturnData as String] = true
  query[kSecMatchLimit as String] = kSecMatchLimitOne
  var item: CFTypeRef?
  guard SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess else { return nil }
  return item as? Data
}

private func storeRepresentation(_ alias: String, _ representation: Data) throws {
  SecItemDelete(keychainQuery(alias) as CFDictionary)
  var query = keychainQuery(alias)
  query[kSecValueData as String] = representation
  query[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
  let status = SecItemAdd(query as CFDictionary, nil)
  guard status == errSecSuccess else {
    throw DeviceKeyKeychainException("The key handle could not be stored, status \(status)")
  }
}

internal final class DeviceKeyUnsupportedException: GenericException<String> {
  override var reason: String { param }
}

internal final class DeviceKeyMissingException: GenericException<String> {
  override var reason: String { "No static key is stored under \(param)" }
}

internal final class DeviceKeyPeerException: GenericException<String> {
  override var reason: String { param }
}

internal final class DeviceKeyKeychainException: GenericException<String> {
  override var reason: String { param }
}

public class DomovoiDeviceKeyModule: Module {
  public func definition() -> ModuleDefinition {
    Name("DomovoiDeviceKey")

    Function("isSupported") { SecureEnclave.isAvailable }

    AsyncFunction("createStaticKey") { (alias: String, requireHardware: Bool) -> [String: Any] in
      if !SecureEnclave.isAvailable && requireHardware {
        throw DeviceKeyUnsupportedException("This device has no Secure Enclave")
      }
      if SecureEnclave.isAvailable {
        let key = try SecureEnclave.P256.KeyAgreement.PrivateKey()
        try storeRepresentation(alias, key.dataRepresentation)
        return report(key.publicKey, "secure-enclave")
      }
      let key = P256.KeyAgreement.PrivateKey()
      try storeRepresentation(alias, key.rawRepresentation)
      return report(key.publicKey, "software")
    }

    AsyncFunction("getStaticKey") { (alias: String) -> [String: Any]? in
      guard let representation = loadRepresentation(alias) else { return nil }
      if SecureEnclave.isAvailable,
        let key = try? SecureEnclave.P256.KeyAgreement.PrivateKey(dataRepresentation: representation) {
        return report(key.publicKey, "secure-enclave")
      }
      guard let key = try? P256.KeyAgreement.PrivateKey(rawRepresentation: representation) else {
        return nil
      }
      return report(key.publicKey, "software")
    }

    AsyncFunction("agree") { (alias: String, peerPublicKey: String) -> String in
      guard let representation = loadRepresentation(alias) else {
        throw DeviceKeyMissingException(alias)
      }
      let peer = try decodePeer(peerPublicKey)
      let secret: SharedSecret
      if SecureEnclave.isAvailable,
        let key = try? SecureEnclave.P256.KeyAgreement.PrivateKey(dataRepresentation: representation) {
        secret = try key.sharedSecretFromKeyAgreement(with: peer)
      } else {
        let key = try P256.KeyAgreement.PrivateKey(rawRepresentation: representation)
        secret = try key.sharedSecretFromKeyAgreement(with: peer)
      }
      return secret.withUnsafeBytes { Data($0).base64EncodedString() }
    }

    AsyncFunction("deleteStaticKey") { (alias: String) -> Bool in
      let existed = loadRepresentation(alias) != nil
      SecItemDelete(keychainQuery(alias) as CFDictionary)
      return existed
    }

    Function("randomBytes") { (count: Int) -> String in
      if count <= 0 || count > 1024 {
        throw DeviceKeyPeerException("Ask for 1 to 1024 bytes")
      }
      var bytes = [UInt8](repeating: 0, count: count)
      let status = SecRandomCopyBytes(kSecRandomDefault, count, &bytes)
      guard status == errSecSuccess else {
        throw DeviceKeyKeychainException("The system random source failed, status \(status)")
      }
      return Data(bytes).base64EncodedString()
    }
  }

  private func report(_ publicKey: P256.KeyAgreement.PublicKey, _ level: String) -> [String: Any] {
    ["publicKey": publicKey.x963Representation.base64EncodedString(), "securityLevel": level]
  }

  private func decodePeer(_ peerPublicKey: String) throws -> P256.KeyAgreement.PublicKey {
    guard let bytes = Data(base64Encoded: peerPublicKey) else {
      throw DeviceKeyPeerException("The peer public key is not base64")
    }
    guard bytes.count == uncompressedPointBytes, bytes.first == 4 else {
      throw DeviceKeyPeerException("The peer public key must be a 65 byte uncompressed P-256 point")
    }
    do {
      return try P256.KeyAgreement.PublicKey(x963Representation: bytes)
    } catch {
      throw DeviceKeyPeerException("The peer public key is not a point on P-256")
    }
  }
}
